### DISCUSSION TITLE
What's the best way to make a mission filter for "anywhere except the source"?

### DISCUSSION THREAD

**Discussion by Zitchas** [endless-sky/endless-sky#7786] [Balance, Mechanics, and Interactions]
Specifically, looking to figure out how to make random jobs that can legitimately have another planet in the same system as a destination, while ensuring that it never has the starting planet as a destination.

This would be for random job board jobs that could be to other planets in the same system or adjacent systems. Short haul transportation. The outer limit is fairly easy to do with our filters, it is the min limit that I've having trouble figuring out.

Looking at the Earth-Luna-Mars in-system missions, and all the ones I've looked at use hard-coded lists for locations instead of general filters, which wouldn't work for what I'm trying to do. Well, it would work, just it'd make for an awfully huge list job.

**Comment by quyykk** [comment]
I'm pretty sure that's not possible. Sounds like a feature request :smile:

**Reply by Zitchas** [reply]
Alright. Thanks!

**Comment by vitalchip** [comment]
Maybe this says something about my background/personality but first thought..

If there's an Earth/Mars job and I park a ship on each, will changing flagships in the info screen complete the mission?

**Reply by quyykk** [reply]
Is it possible to change planets like this? I hope not. So the answer is hopefully no.

**Reply by vitalchip** [reply]
Yes, it is.  If you park a ship on Earth, depart in your flagship to Mars and land, unpark the ship on Earth and make it the flagship you'll depart from Mars in the other ship.

**Reply by quyykk** [reply]
bruh

**Reply by warp-core** [reply]
It doesn't complete missions (I just tested it), you just depart from a different planet, but until you depart, you still apear to be on the planet you landed on.

**Reply by vitalchip** [reply]
That's fine then, well, at least for this!

Still don't like the fact you can outfit ships which are on a planet with no outfitter just as long as the flagship is in system and landed somewhere there is one.  It's a small thing and 'player convenience' but sooner or later it'll work against something..

**Reply by quyykk** [reply]
Seems like we shouldn't group ships in the same system together.

**Reply by vitalchip** [reply]
It was brought up a while back and changing it was rejected because it's a player convenience. A bit like all ships landing with the flagship provided they are in system. It's a player convenience until it's a player exploit.

**Reply by Zitchas** [reply]
Yeah, not a fan of that, but ah well. Thanks for testing! As far as that goes, not sure you should be able to switch flagships to a planet that is parked on a different planet...

**Reply by vitalchip** [reply]
It does still create a bit of an issue, for example I would like freight runs between Earth and Neptune. They'd pay well because Neptune is really far away, so far that it takes longer to get there than jumping to the next system.

However..  I can put a load of cargo into some haulers and then just zoom ahead in the smallest fastest ship I can get, land at Neptune while they're still only 10% of the way there and they're magic'd in alongside me.

Worse yet, I can park a cargo ship at Neptune and one at Earth, accept the mission at Earth, switch flagships and when I depart Neptune (post switch) all my cargo (from Earth supposedly) gets loaded into the new flagship. A quick 180 and land. Job done with none of that silly travelling through space. 

If there were repeating jobs in both directions it's basically a land/switch/depart cargo teleporter.

**Reply by Zitchas** [reply]
Ah, fleet mechanics. Breaking stuff since the day we could switch flagship.

**Comment by vitalchip** [comment]
Allowing flagship changes only when the ships are on the same planet is simple enough but we'd also need some player feedback, to at least give them a clue why it might not be changing. I think just letting them know which planet their ships are on would be enough.

I haven't looked but I'm guessing ShopPanel can be changed similarly to prevent ships appearing in the sidebar if they're not on the same planet as the player.

<img width="724" alt="Screenshot 2022-12-06 at 16 21 03" src="https://user-images.githubusercontent.com/65342577/205971738-5d174738-8aa5-480d-accf-0513a87204cd.png">

<img width="1025" alt="Screenshot 2022-12-06 at 16 29 20" src="https://user-images.githubusercontent.com/65342577/205971786-869ab03d-b19b-4de6-ae00-5c7d211a3bb0.png">

<img width="1033" alt="Screenshot 2022-12-06 at 16 29 29" src="https://user-images.githubusercontent.com/65342577/205971793-ea9a9863-53c8-4be4-95f6-1346694dd041.png">

**Reply by Zitchas** [reply]
I think that'd be a valuable addition, since multi-planet systems are probably going to be increasing in number...

**Comment by adegans** [comment]
@Zitchas 
Something like this should work, as implied by the manual.
https://github.com/endless-sky/endless-sky/wiki/CreatingMissions

```
	source
		system "Republic" "Free Worlds" "Syndicate" "Neutral" "Independent"
	destination
		distance 0
		not <origin>
```

**Reply by warp-core** [reply]
That won't work. `<origin>` is a text replacement. This is not a text field.
Text replacements work in: the `name`, `description`, `blocked` messages, `conversation`s, and `dialog`s in that mission.

**Reply by adegans** [reply]
Oh, duh! 🤪
But something to this extend then, but with a correct value.
Or simply "distance 0". But then logically a filter should be added to detect if there are >1 planets/locations in system where a player can land. And I didn't find any of such logic in the manual.
